File: src/main/java/com/illuzionzstudios/mist/ui/button/type/ReturnBackButton.java
package com.illuzionzstudios.mist.ui.button.type;

import com.cryptomorin.xseries.XMaterial;
import com.illuzionzstudios.mist.item.ItemCreator;
import com.illuzionzstudios.mist.ui.UserInterface;
import com.illuzionzstudios.mist.ui.button.Button;
import lombok.*;
import org.bukkit.inventory.ItemStack;

import java.util.Arrays;
import java.util.List;

/**
 * A button that returns to a previous/parent {@link UserInterface}
 */
@RequiredArgsConstructor
@AllArgsConstructor
public final class ReturnBackButton extends Button {

    /**
     * Material for this button
     */
    @Getter
    @Setter
    private static XMaterial material = XMaterial.OAK_DOOR;

    /**
     * The title of this button
     */
    @Getter
    @Setter
    private static String title = "&4&lReturn";

    /**
     * The lore of this button
     */
    @Getter
    @Setter
    private static List<String> lore = Arrays.asList("", "Return back.");

    /**
     * The parent {@link UserInterface}
     */
    @NonNull
    private final UserInterface parentInterface;

    /**
     * Make a new instance of the {@link UserInterface} when showing
     */
    private boolean makeNewInstance = false;

    @Override
    public ItemStack getItem() {
        return ItemCreator.of(material).name(title).lores(lore).build().makeUIItem();
    }

    /**
     * Open the parent interface
     */
    @Override
    public ButtonListener getListener() {
        return ((player, ui, type, event) -> {
            // When clicking don't move items
            event.setCancelled(true);
            if (makeNewInstance)
                parentInterface.newInstance().show(player);
            else
                parentInterface.show(player);
        });
    }
}

File: src/main/kotlin/com/illuzionzstudios/mist/ui/button/type/ReturnBackButton.kt
package com.illuzionzstudios.mist.ui.button.type

import com.cryptomorin.xseries.XMaterial
import com.illuzionzstudios.mist.item.ItemCreator
import com.illuzionzstudios.mist.ui.UserInterface
import com.illuzionzstudios.mist.ui.button.Button
import lombok.*
import org.bukkit.entity.Player
import org.bukkit.event.inventory.ClickType
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.inventory.ItemStack
import java.util.*

/**
 * A button that returns to a previous/parent [UserInterface]
 */
@RequiredArgsConstructor
@AllArgsConstructor
class ReturnBackButton : Button() {
    /**
     * The parent [UserInterface]
     */
    private val parentInterface: UserInterface = null

    /**
     * Make a new instance of the [UserInterface] when showing
     */
    private val makeNewInstance = false
    override val item: ItemStack?
        get() = ItemCreator.Companion.of(material).name(title).lores(lore).build()
            .makeUIItem()// When clicking don't move items

    /**
     * Open the parent interface
     */
    override val listener: ButtonListener
        get() = ButtonListener { player: Player, ui: UserInterface?, type: ClickType?, event: InventoryClickEvent ->
            // When clicking don't move items
            event.isCancelled = true
            if (makeNewInstance) parentInterface.newInstance().show(player) else parentInterface.show(player)
        }

    companion object {
        /**
         * Material for this button
         */
        @Getter
        @Setter
        private val material = XMaterial.OAK_DOOR

        /**
         * The title of this button
         */
        @Getter
        @Setter
        private val title = "&4&lReturn"

        /**
         * The lore of this button
         */
        @Getter
        @Setter
        private val lore = Arrays.asList("", "Return back.")
    }
}